include/wil/network.h

```diff
@@ -0,0 +1,1338 @@
+    inline ::wil::network::socket_address map_dual_mode_4to6(const ::wil::network::socket_address& ipv4_address) WI_NOEXCEPT
+    {
+        constexpr IN6_ADDR ipv4MappedPrefix{{IN6ADDR_V4MAPPEDPREFIX_INIT}};
+
+        ::wil::network::socket_address return_ipv6_address{&ipv4MappedPrefix, ipv4_address.port()};
+
+        auto* const return_in6_addr{return_ipv6_address.in6_addr()};
+        const auto* const ipv4_inaddr{ipv4_address.in_addr()};
+        return_in6_addr->u.Byte[12] = ipv4_inaddr->S_un.S_un_b.s_b1;
+        return_in6_addr->u.Byte[13] = ipv4_inaddr->S_un.S_un_b.s_b2;
+        return_in6_addr->u.Byte[14] = ipv4_inaddr->S_un.S_un_b.s_b3;
+        return_in6_addr->u.Byte[15] = ipv4_inaddr->S_un.S_un_b.s_b4;
+
+        return return_ipv6_address;
+    }
+
+    // non-member swap
+    inline void swap(::wil::network::socket_address& lhs, ::wil::network::socket_address& rhs) WI_NOEXCEPT
+    {
+        lhs.swap(rhs);
+    }
+
+    // class addr_info encapsulates the ADDRINFO-related structures returned from the socket functions
+    // getaddrinfo, GetAddrInfoW, GetAddrInfoWEx
+    // iterator semantics are supported to safely access these addresses
+    // template T supports pointers to the 3 address structures: ADDRINFOA*, ADDRINFOW*, ADDRINFOEXW*
+    template <typename T>
+    class addr_info_iterator_t
+    {
+    public:
+        // defining iterator_traits allows STL <algorithm> functions to be used with this iterator class.
+        // Notice this is a forward_iterator
+        // - does not support random-access (e.g. vector::iterator)
+        // - does not support bidirectional access (e.g. list::iterator)
+#if WIL_USE_STL || defined(WIL_DOXYGEN)
+        using iterator_category = ::std::forward_iterator_tag;
+#endif
+        using value_type = ::wil::network::socket_address;
+        using difference_type = size_t;
+        using distance_type = size_t;
+        using pointer = ::wil::network::socket_address*;
+        using reference = ::wil::network::socket_address&;
+
+        constexpr addr_info_iterator_t() WI_NOEXCEPT = default;
+
+        explicit addr_info_iterator_t(const T* addrinfo) WI_NOEXCEPT :
+            // must const cast so we can re-use this pointer as we walk the list
+            m_addrinfo_ptr(const_cast<T*>(addrinfo))
+        {
+            if (m_addrinfo_ptr)
+            {
+                m_socket_address.set_sockaddr(m_addrinfo_ptr->ai_addr, m_addrinfo_ptr->ai_addrlen);
+            }
+        }
+
+        ~addr_info_iterator_t() WI_NOEXCEPT = default;
+        addr_info_iterator_t(const addr_info_iterator_t&) WI_NOEXCEPT = default;
+        addr_info_iterator_t& operator=(const addr_info_iterator_t&) WI_NOEXCEPT = default;
+        addr_info_iterator_t(addr_info_iterator_t&&) WI_NOEXCEPT = default;
+        addr_info_iterator_t& operator=(addr_info_iterator_t&&) WI_NOEXCEPT = default;
+
+        const ::wil::network::socket_address& operator*() const WI_NOEXCEPT

```

Comment:
```suggestion
        }

        const ::wil::network::socket_address& operator*() const WI_NOEXCEPT
```
Already the default

Reply:
move operators will automatically be created too?

Reply:
oh, it really doesn't matter for this trivial class.

Reply:
I'll link some resources later once I find it again. The TL;DR is that the specification for which constructors/assignment operators are defined for you when is a _little_ complex, however if you follow the very general "rule of five" (where you either specify all or none of the copy/move constructors, copy/move assignment operators, and the destructor), you'll get "the right thing" 99.9% of the time.

The downside to "well I'll just always specify them just to be safe" is that `= default` is not the same as implicit and this can affect things like whether or not the type is trivial, which can have performance implications.

Reply:
As promised...

Here's an overview of "the rule of five" (and "the rule of zero"): https://www.youtube.com/watch?v=juAZDfsaMvY

Here's the chart from that talk about which functions are defined and when, which is a good resource (but is not necessary if you follow the rule of five/zero):
![image](https://github.com/user-attachments/assets/9bce2f50-8a43-4a2c-a193-84c34a863679)
And the "blog post" that this image comes from: https://howardhinnant.github.io/classdecl.html

From the Core Guidelines:
[C.20 The rule of zero](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-zero)
[C.21 The rule of five](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-five)